[Term Entry] Dart List - Added list.md which describes Dart Lists (#4282)
Codecademy/docs: content/dart/concepts/list/list.md

<!-- I think will be worth linking to other docs pages like 'strings' when they have been completed / submitted. -->

## Syntax

In Dart, you are able to create a list utilising the `List` class, which also uses square brackets (`[]`) and commas (`,`). The square brackets are used to define the data, whilst commas are used to separate the individual items.

```dart
// Create a list of strings
List favouriteFruits = ['Apple', 'Banana', 'Orange'];

// Create a list of integers
List oddNumbers = [1, 3, 5, 7, 9];

// Create a list of lists (nested list)
List nestedList = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
];
```

In the above example, there are three separate lists:
- A list of strings
- A list of integers
- A nested list

Each list includes the same data type. However, it is possible to create lists using a mix of data types.

```dart
// Create a list with mixed data types
List mixedList = [1, 'two', 3, true];
```

## Different List Types

> There are two types of lists in Dart; fixed length or growable. A fixed length list has a specified size which cannot be changed after initialistation, whereas a growable list can have data added or removed.

Reviewer: ```suggestion
There are two types of lists in Dart, fixed-length and growable. A fixed-length list has a specified size that cannot be changed after initialization. When creating a fixed-length list, it starts empty and needs items to be added to it. It is useful when the number of items in a list is known in advance. The following code block demonstrates the initialization of a fixed-length list:
```

```dart
// Example of a fixed length list
List fixedLengthList = List(3);
```

> 

Reviewer: ```suggestion

>**Note:** If more than the specified number of items will result in an error.

```

When creating a fixed length list, it starts off empty and needs items added to it. Adding too many items will result in an error. Fixed length lists are useful when you know in advance the number of items you need and want to enforce that size constraint. 

```dart
// Example of a growable list
List growableList = ['Lion', 'Tiger', 'Leopard'];
```

> 
> When creating a growable list, we are able to start with a set of data. We can then add additional data or remove data without causing an error. Growable lists provide flexibility and grow or shrink dynamically based on requirements.

Reviewer: ```suggestion
```

  - 'Web Development'
Tags:
  - 'Lists'
  - 'Arrays'
  - 'Objects'
  - 'Data Structures'
  - 'Data Types'
CatalogContent:
  - 'learn-dart'
  - 'paths/computer-science'
---

A **list** provides a way to organize and store a collection of items. Simply, lists act as containers that can hold different pieces of data, such as strings, integers, boolean values, or even other lists. Lists are a great way to organize related data in an easy-to-manipulate format.
<!-- I think will be worth linking to other docs pages like 'strings' when they have been completed/submitted. -->

## Syntax

In Dart, there are two ways to create a list, there is the more prominently used method of `var` and the lesser used `List` class, both use square brackets (`[]`) and commas (`,`). The square brackets are used to define the data, whilst commas are used to separate the individual items.

First, an example of utilizing `var` to create a list:

```dart
// Create a list of strings
var favouriteFruits = ['Apple', 'Banana', 'Orange'];

// Create a list of integers
var oddNumbers = [1, 3, 5, 7, 9];

// Create a list of lists (nested list)
var nestedList = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
];
```

> Then, an example of utilizing `List` to create a list:

Reviewer 2: ```suggestion
Then, an example of utilizing `List` to create a list:

```

```dart
// Create a list of strings
List favouriteFruits = ['Apple', 'Banana', 'Orange'];

// Create a list of integers
List oddNumbers = [1, 3, 5, 7, 9];

// Create a list of lists (nested list)
List nestedList = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
];
```

>**Note:** As `var` is the predominantly used method to create lists in Dart, the rest of the examples will use this moving forward.

In the above examples, there are three separate lists:
- A list of strings
- A list of integers

> - A nested list

Reviewer 2: You can add this before the example b=code block. It would be better for the users to understand the code block.

Each list includes the same data type. However, creating lists using a mix of data types is possible.

```dart
// Create a list with mixed data types
var mixedList = [1, 'two', 3, true];
```

## Different List Types

There are two types of lists in Dart, fixed-length and growable. A fixed-length list has a specified size that cannot be changed after initialization. When creating a fixed-length list, it starts empty and needs items to be added to it. It is useful when the number of items in a list is known in advance. The following code block demonstrates the initialization of a fixed-length list:

```dart
// Example of a fixed length list
var fixedLengthList = List(3);
```

> >**Note:** If more than the specified number of items are added, it will result in an error.

Reviewer 2: ```suggestion
> **Note:** If more than the specified number of items are added, it will result in an error.
```

When creating a fixed-length list, it starts empty and needs items added to it. Fixed-length lists are useful when a known limit on list items is required ahead of time. This enforces the list to stay at the specified limit.

> 

Reviewer 2: Add a line describing the growable list example.

```dart
// Example of a growable list
var growableList = ['Lion', 'Tiger', 'Leopard'];
```

## List Methods

Lists are very versatile sets of data and Dart provides ways to manage the data held within the lists called methods. These methods include things like adding new data and removing data.

### Adding Data

Adding items of data to lists utilizing the following methods:
- `(listname).add(item);` - This adds an item to the end of the list
- `(listname).addAll(items);` - This adds multiple items to the end of the list

```dart
// Adding items to a list
var ages = [21, 23, 34, 19, 45];

ages.add(42); // Adds 42 to the end of the list
ages.addAll(42, 37, 62); // Adds 42, 37 and 62 to the end of the list
```

It is also possible to add items at specific points within a list. This can be done by calling the list and specifying where the item should be added.

```dart
// Adding an item to a specific position within a list
var ages = [21, 23, 34, 19, 45];

ages[0] = 87; // Adds 87 to the first position of the list
```

> This comes in handy with fixed length lists:

Reviewer 2: ```suggestion

This comes in handy with fixed-length lists:
```
Give one line space before and after a code block.

```dart
var fixedLengthList = List(3);

fixedLengthList[0] = 1; // Sets the 1st item in the list as 1
fixedLengthList[1] = 2; // Sets the 2nd item in the list as 2
fixedLengthList[2] = 3; // Sets the 3rd item in the list as 3
```

### Removing Data

Removing items of data from lists utilizes the following methods:
- `(listname).remove(item);` - This removes the first occurance of the specified item
- `(listname).removeAt(1);` - This removes the item at the specified index
- `(listname).removeLast();` - This removes the last item in the list

```dart
var favouriteColors = ['red', 'red', 'orange', 'yellow', 'green'];

favouriteColors.remove(red); // Removes the first 'red' in the list
favouriteColors.removeAt(1); // Removes the second item in the list
favouriteColors.removeLast(); // Removes the last item from the list
```

## Displaying List Data

> To display data held within a list the `print` method is used. This can be used to display the whole list, or specific items within the list.

Reviewer 2: ```suggestion

To display data held within a list the `print` method is used. This can be used to display the whole list, or specific items within the list.
```

```dart
// Displaying a whole list
var flowers = ['Rose', 'Lily', 'Sunflower', 'Tulip', 'Daisy'];

print(flowers); // Displays ['Rose', 'Lily', 'Sunflower', 'Tulip', 'Daisy'] in the console.

// Displaying a specific item from a list
var musicalInstruments = ['Piano', 'Guitar', 'Violin', 'Trumpet'];

> print(musicalInstruments[2]); // Displays Violin in the console

Reviewer 2: Wrap this code under the main() function so that it will be executable as it is.

var ages = [21, 23, 34, 19, 45];

ages[0] = 87; // Adds 87 to the first position of the list
```

This comes in handy with fixed-length lists:

```dart
var fixedLengthList = List(3);

fixedLengthList[0] = 1; // Sets the 1st item in the list as 1
fixedLengthList[1] = 2; // Sets the 2nd item in the list as 2
fixedLengthList[2] = 3; // Sets the 3rd item in the list as 3
```

### Removing Data

Removing items of data from lists utilizes the following methods:

- `(listname).remove(item);` - This removes the first occurance of the specified item
- `(listname).removeAt(1);` - This removes the item at the specified index
- `(listname).removeLast();` - This removes the last item in the list

```dart
var favouriteColors = ['red', 'red', 'orange', 'yellow', 'green'];

favouriteColors.remove(red); // Removes the first 'red' in the list
favouriteColors.removeAt(1); // Removes the second item in the list
favouriteColors.removeLast(); // Removes the last item from the list
```

## Displaying List Data

To display data held within a list the `print` method is used. This can be used to display the whole list, or specific items within the list.

```dart

> /void main() {

Reviewer 2: ```suggestion
void main() {
```
